### ISSUE TITLE
[FATAL BUG REPORT] with torch.no_grad(): can not close the gradient of lora matrix

### ISSUE THREAD

**Issue by bjzhb666** [microsoft/LoRA#185]
Thanks for your great great work. I found a serious bug recently. 

After we define a model using Lora layers, we get a model with lora layers instead of nn.Linear.

But when we use `with torch.no_grad():`, I find the this operation can not close the gradient of lora.

Here is my demo code:
```
# define a backbone with lora layers
BACKBONE=lora_model()

with torch.no_grad():
    # print the gradient of the model
    for n,p in BACKBONE.named_parameters():
          if p.grad is not None:
              print(f"grad of {n}: {p.grad.norm()}")

```
**I found that not all the layers closed the gradient.**

Here is an output example.
![Image](https://github.com/user-attachments/assets/d9348fb3-6806-43da-b610-46dfd8a9d3b8)

**This error is fatal because when we conduct eval during the training phase, the gradient will be influenced.** 

Hope you can fix it!!!

**Comment by bjzhb666** [comment]
How can I close the gradient of these lora modules? Please help me!

**Comment by emi-dm** [comment]
Please, provide all the code in order to replicate the issue @bjzhb666

**Comment by emi-dm** [comment]
Here is an example of how to use it: 

```
import torch
from torch import nn
import torchvision
from torchvision import datasets, models, transforms
from utils import print_trainable_params
import loralib as lora

#model_ft = models.resnet18(weights=models.ResNet18_Weights.IMAGENET1K_V1)
# torch.save(model_ft.state_dict(), 'resnet18.pth')


model_ft = models.resnet18(weights=None)
model_ft.load_state_dict(torch.load('resnet18.pth', weights_only=True), strict=False)
original_weights = model_ft.layer1[0].conv1.weight.clone()
original_linear_weights = model_ft.fc.weight.clone()



# Replace the first convolutional layer with a LoRA convolutional layer
lora_model  = models.resnet18(weights=None)
lora_model.layer1[0].conv1 = lora.Conv2d(64, 64, kernel_size=3, r=8, stride=(1,1), padding=(1,1), bias=False)
lora_model.fc = lora.Linear(512, 1000, r=8, bias=True)
lora_model.load_state_dict(torch.load('resnet18.pth', weights_only=True), strict=False)

lora_layer_weights = lora_model.layer1[0].conv1.conv.weight.clone()
lora_linear_weights = lora_model.fc.weight.clone()
lora.mark_only_lora_as_trainable(lora_model)

# Check if the weights are equal
assert torch.allclose(original_linear_weights, lora_linear_weights), "Linear Weights are not equal"
assert torch.allclose(original_weights, lora_layer_weights), "Conv Weights are not equal"

print(lora_model.layer1[0].conv1.lora_A.shape)
print(lora_model.layer1[0].conv1.lora_B.shape)

print_trainable_params(lora_model)

with torch.no_grad():
    for n, p in lora_model.named_parameters():
        if p.grad is not None:
            print(f"grad of {n}: {p.grad.norm()}")
```

**Comment by bjzhb666** [comment]
I am using a Transformer model, let me find an example. Please wait me for a while.

**Comment by bjzhb666** [comment]
Sorry for posting this issue. I find the gradient does not change as you mentioned. 
But when we use model.eval(), since we add LORA's weight to the original model, **this will affect the numerical precision of the decimal number, which will affect our results.**

**Comment by bjzhb666** [comment]
Sorry for bothering you, it not a bug. Many thanks to your reply
